src/EspressoSGXTEEVerifier.sol

```diff
@@ -0,0 +1,179 @@
+        registeredEnclaveHash[enclaveHash] = true;
+        registeredEnclaveSigner[enclaveSigner] = true;
+    }
+
+    /*
+        @notice Verify a quote from the TEE and attest on-chain
+        The verification is considered successful if the function does not revert.
+        @param rawQuote The quote from the TEE
+        @param reportDataHash The hash of the report data
+    */
+    function verify(bytes calldata rawQuote, bytes32 reportDataHash)
+        public
+        view
+        returns (EnclaveReport memory)
+    {
+        // Parse the header
+        Header memory header = parseQuoteHeader(rawQuote);
+
+        // Currently only version 3 is supported
+        if (header.version != 3) {
+            revert InvalidHeaderVersion();
+        }
+
+        // Verify the quote
+        (bool success,) = quoteVerifier.verifyQuote(header, rawQuote);
+        if (!success) {
+            revert InvalidQuote();
+        }
+
+        // Parse enclave quote
+        uint256 lastIndex = HEADER_LENGTH + ENCLAVE_REPORT_LENGTH;
+        EnclaveReport memory localReport;
+        (success, localReport) = parseEnclaveReport(rawQuote[HEADER_LENGTH:lastIndex]);
+        if (!success) {
+            revert FailedToParseEnclaveReport();
+        }
+
+        // Check that mrEnclave match
+        if (!registeredEnclaveHash[localReport.mrEnclave]) {
+            revert InvalidEnclaveHash();
+        }
+
+        if (!registeredEnclaveSigner[localReport.mrSigner]) {
+            revert InvalidEnclaveSigner();
+        }
+
+        //  Verify that the reportDataHash if the hash signed by the TEE
+        // We do not check the signature because `quoteVerifier.verifyQuote` already does that
+        if (reportDataHash != bytes32(localReport.reportData.substring(0, 32))) {
+            revert InvalidReportDataHash();
+        }
+
+        return localReport;
+    }
+
+    /*
+        @notice Register a new signer by verifying a quote from the TEE
+        @param attestation The attestation from the TEE
+        @param data which the TEE has attested to
+    */
+    function registerSigner(bytes calldata attestation, bytes calldata data) external {
+        // Check that the data length is 20 bytes because an address is 20 bytes
+        if (data.length != 20) {
+            revert InvalidDataLength();
+        }
+        bytes32 paddedSignerAddress = keccak256(data);
+        // Convert data to address
+        EnclaveReport memory localReport = verify(attestation, paddedSignerAddress);
+
+        if (localReport.reportData.length < 20) {
+            revert ReportDataTooShort();
+        }
+
+        address signer = address(uint160(bytes20(data[:20])));
+
+        // Check if the extracted address is valid
+        if (signer == address(0)) {
+            revert InvalidSignerAddress(); // Custom revert if the address is invalid
+        }
+        // Mark the signer as registered
+        registeredSigners[signer] = true;
```

Comment:
you can add a check that the signer isn't already registered to prevent unneccessary storage updates

Reply:
true, will do that!

Reply:
https://github.com/EspressoSystems/espresso-tee-contracts/pull/2/commits/a2bf9e83c0441bd275e4cb10f5bb23b8d440a8ad